backends/rapidpro/msg.go

```diff
@@ -0,0 +1,271 @@
+package rapidpro
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/nyaruka/courier"
+	"github.com/nyaruka/courier/utils"
+	filetype "gopkg.in/h2non/filetype.v1"
+)
+
+// MsgDirection is the direction of a message
+type MsgDirection string
+
+// Possible values for MsgDirection
+const (
+	MsgIncoming     MsgDirection = "I"
+	MsgOutgoing     MsgDirection = "O"
+	NilMsgDirection MsgDirection = ""
+)
+
+// MsgPriority is the priority of our message
+type MsgPriority int
+
+// Possible values for MsgPriority
+const (
+	BulkPriority    MsgPriority = 100
+	DefaultPriority MsgPriority = 500
+	HighPriority    MsgPriority = 1000
+)
+
+// MsgVisibility is the visibility of a message
+type MsgVisibility string
+
+// Possible values for MsgVisibility
+const (
+	MsgVisible  MsgVisibility = "V"
+	MsgDeleted  MsgVisibility = "D"
+	MsgArchived MsgVisibility = "A"
+)
+
+// WriteMsg creates a message given the passed in arguments, returning the uuid of the created message
+func writeMsg(b *backend, msg *courier.Msg) error {
+	m := newMsgFromMsg(msg)
+
+	// if we have media, go download it to S3
+	for i, attachment := range msg.Attachments {
+		if strings.HasPrefix(attachment, "http") {
+			url, err := downloadMediaToS3(b, msg.UUID, attachment)
+			if err != nil {
+				return err
+			}
+			msg.Attachments[i] = url
+		}
+	}
+
+	// try to write it our db
+	err := writeMsgToDB(b, m)
+
+	// fail? spool for later
+	if err != nil {
+		return courier.WriteToSpool(b.config.SpoolDir, "msgs", m)
+	}
+
+	// finally try to add this message to our handling queue
+	err = addToHandleQueue(b, m)
+
+	// TODO: spool backdown for failure to add to redis
+	return err
+}
+
+func newMsgFromMsg(m *courier.Msg) *Msg {
```

Comment:
Different backends don't interact so I'd be less worried about two different `DbMsg` names, but your call.